core/domain/exp_jobs_one_off.py

```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    in the exploration. For valid explorations, they are upgraded to the
+    appropriate new interaction.
+
+    This validation is done by the validator functions present in schema_utils.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        is_valid_algebraic_expression = schema_utils.get_validator(
+            'is_valid_algebraic_expression')
+        is_valid_numeric_expression = schema_utils.get_validator(
+            'is_valid_numeric_expression')
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
+
+        if not item.deleted:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+            for state_name, state in exploration.states.items():
+                if state.interaction.id == 'MathExpressionInput':
+                    new_answer_groups = []
+                    types_of_inputs = set()
+                    for group in state.interaction.answer_groups:
+                        new_answer_group = copy.deepcopy(group)
+                        for rule_spec in new_answer_group.rule_specs:
+                            rule_input = ltt.latex_to_text(
+                                rule_spec.inputs['x'])
+
+                            rule_input = _clean_math_expression(rule_input)
+
+                            type_of_input = _TYPE_INVALID_EXPRESSION
+                            if is_valid_algebraic_expression(rule_input):
+                                type_of_input = _TYPE_VALID_ALGEBRAIC_EXPRESSION
+                            elif is_valid_numeric_expression(rule_input):
+                                type_of_input = _TYPE_VALID_NUMERIC_EXPRESSION
+                            elif is_valid_math_equation(rule_input):
+                                type_of_input = _TYPE_VALID_MATH_EQUATION
+
+                            types_of_inputs.add(type_of_input)
+
+                            if type_of_input != _TYPE_INVALID_EXPRESSION:
+                                rule_spec.inputs['x'] = rule_input
+                                if type_of_input == _TYPE_VALID_MATH_EQUATION:
+                                    rule_spec.inputs['y'] = 'both'
+                                rule_spec.rule_type = 'MatchesExactlyWith'
 
                             output_values = '%s %s: %s' % (
                                 item.id, state_name, rule_input)
 
-                            yield (validity, output_values.encode('utf-8'))
+                            yield (type_of_input, output_values.encode('utf-8'))
+
+                        new_answer_groups.append(new_answer_group.to_dict())
+
+                    if _TYPE_INVALID_EXPRESSION in types_of_inputs:
+                        yield ('ERROR', (
+                            'There are some invalid inputs that need to be '
```

Comment:
This is not a helpful message. It does not help with debugging.

Reply:
The invalid inputs are displayed in the output with the key 'Invalid'. I've added this in the error message as well to avoid confusion.